consensus/src/transaction_shuffler/fairness/mod.rs

```diff
@@ -0,0 +1,165 @@
+
+pub(crate) mod conflict_key;
+mod conflict_zone;
+mod pending_zone;
+mod selection_tracker;
+
+#[cfg(test)]
+mod tests;
+
+type TxnIdx = usize;
+
+const NUM_CONFLICT_ZONES: usize = 3;
+
+struct FairnessShuffler {
+    sender_conflict_window_size: usize,
+    module_conflict_window_size: usize,
+    entry_fun_conflict_window_size: usize,
+}
+
+impl TransactionShuffler for FairnessShuffler {
+    fn shuffle(&self, txns: Vec<SignedTransaction>) -> Vec<SignedTransaction> {
+        let conflict_key_registries = ConflictKeyRegistry::build_registries(&txns);
+        let order = FairnessShufflerImpl::new(self, &conflict_key_registries).shuffle();
+        reorder(txns, order)
+    }
+}
+
+fn reorder(txns: Vec<SignedTransaction>, order: Vec<TxnIdx>) -> Vec<SignedTransaction> {
+    assert_eq!(txns.len(), order.len());
+    order.into_iter().map(|idx| txns[idx].clone()).collect()
+}
+
+struct FairnessShufflerImpl<'a> {
+    conflict_key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+    conflict_zones: [ConflictZone<'a>; NUM_CONFLICT_ZONES],
+    pending_zones: [PendingZone<'a>; NUM_CONFLICT_ZONES],
+    selected_order: Vec<TxnIdx>,
+    selection_tracker: SelectionTracker,
+}
+
+impl<'a> FairnessShufflerImpl<'a> {
+    pub fn new(
+        shuffler: &FairnessShuffler,
+        conflict_key_registries: &'a [ConflictKeyRegistry; NUM_CONFLICT_ZONES],
+    ) -> Self {
+        let num_txns = conflict_key_registries[0].num_txns();
+        assert!(conflict_key_registries
+            .iter()
+            .skip(1)
+            .all(|r| r.num_txns() == num_txns));
+
+        Self {
+            conflict_key_registries,
+            selected_order: Vec::with_capacity(num_txns),
+            selection_tracker: SelectionTracker::new(num_txns),
+            conflict_zones: ConflictZone::build_zones(conflict_key_registries, [
+                shuffler.sender_conflict_window_size,
+                shuffler.module_conflict_window_size,
+                shuffler.entry_fun_conflict_window_size,
+            ]),
+            pending_zones: PendingZone::build_zones(conflict_key_registries),
+        }
+    }
+
+    pub fn shuffle(mut self) -> Vec<TxnIdx> {
+        // First pass, only select transactions with no conflicts in all conflict zones
+        while let Some(txn_idx) = self.selection_tracker.next_unselected() {
+            if !self.is_conflict(txn_idx) && !self.is_hol_blocked(txn_idx) {
+                self.select_and_select_unconflicted(txn_idx, false /* is_pending */)
+            } else {
+                self.append_pending(txn_idx);
+            }
+        }
+
+        // Second pass, select previously pending txns in order,
+        //   with newly un-conflicted txns jumping the line
+        self.selection_tracker.new_pass();
+        while let Some(txn_idx) = self.selection_tracker.next_unselected() {
+            self.select_and_select_unconflicted(txn_idx, true /* is_pending */);
+        }
+
+        self.selected_order
```

Comment:
Can you add a these two tests (window 2 on modules, and other two categories not conflicting):

`moduleA, moduleA, moduleA, moduleB, moduleC, moduleD, moduleE`
reorders into:
`moduleA, moduleB, moduleA, moduleC, moduleA, moduleD, moduleE`

and:

`moduleA, moduleA, moduleA, moduleB, moduleC, moduleD, moduleE, moduleB`
=>
`moduleA, moduleB, moduleA, moduleC, moduleA, moduleD, moduleE, moduleB`

I think current logic will fail on the second one, if I am following it correctly

Reply:
hmm.. I added the test and it works like expected (`test_dominating_module2()`). What particularly do you see sketchy? 

(`select_and_select_unconflicted()` was new, maybe you read a version without it?)

Reply:
I expected that select_and_select_unconflicted will pull second moduleB in once it's dependency gets resolved (i.e. at position 4), even though some other transactions have lower index

Reply:
> I expected that select_and_select_unconflicted will pull second moduleB in once it's dependency gets resolved (i.e. at position 4), even though some other transactions have lower index

It won't, because when the first pass sees Module E, the second Module B hasn't yet been marked pending. 
The test verified that.